text/0055-remove-unspecified.md

```diff
@@ -0,0 +1,225 @@
+Unspecified and unknown entities are both pre-defined "dummy" entities that act like any other entity, as long as they are unused or only used in a trivial way (e.g., `principal == principal`).
+If a policy uses an _unknown entity_ in a non-trivial way, then evaluation will produce a residual.
+If a policy uses an _unspecified entity_ in a non-trivial way, then evaluation will produce an error.
+
+Unspecified entities also lead to two sharp corners in our public API:
+
+1. Omitting a `appliesTo` field in the schema (or setting it to null) means that an action applies only to unspecified entities, while using an empty list means that it applies to no entities. There is no way to say that an action applies to both unspecified entities and other types of entities. For more discussion, see [RFC35](https://github.com/cedar-policy/rfcs/pull/35). (Note: RFC35 has been closed in favor of the current RFC.)
+2. It is unclear how we should create `Request`s that use both unspecified and unknown entities. This has not been an issue so far since partial evaluation and its APIs are experimental, but we are looking to stabilize them soon. Currently the best (although not ideal) proposal is to use a builder pattern for `Request`s (e.g., `request.principal(Some(User::"alice"))` sets the principal field of a request). To set a field to be “unspecified” users should pass in `None` instead of `Some(_)`; to set a field to be unknown users should not call the field constructor at all.
+
+## Detailed design
+
+### Naming specifics
+
+The names `RequestEntity` and `__cedar::Default` are both subject to change.
+
+For the latter, we have agreed that we want to use the `__cedar` prefix (reserved as part of [cedar#557](https://github.com/cedar-policy/cedar/pull/557) in v3.1), but the exact type name is up for debate.
+Options we've considered include:
+
+- `None`
+- `Null`
+- `Ghost`
+- `Dummy`
+- `Empty`
+- `Irrelevant`
+- `Generic`
+
+Here are some options we’ve ruled out:
+
+- `Any`: possibility to confuse the scope condition `principal` with `principal is __cedar::Any`. The first applies to any principal, while the latter applies to only the “unspecified” entity.
+- `Arbitrary`: same issue as “any”
+- `Unspecified`: potential for confusion with partial evaluation “unknown”
+- `Mock`: implies that this entity type should be used for debugging/testing
+
+### `Default` vs. `EntityUid(__cedar::Default::"eid")`
+
+Under the current proposal, there are two ways to pass an unspecified entity into a `Request`: `RequestEntity::Default` and `RequestEntity::EntityUid(__cedar::Default::"eid")`.
+Both are equivalent, although the latter allows the user to assign an particular name ("eid") to the entity, which could be helpful when debugging error messages.
+When using the `Default` variant, Cedar will assign the entity id "principal", "action", or "resource" depending on where the entity is used in the request.
+
```

Comment:
In general, I like this option, but I'm concerned about having two distinct ways to do the same thing.  If I understand correctly, users can construct an _arbitrary_ entity of the `__cedar::Default` and pass it in?  So, I could construct a request that binds both the `principal` and `resource` variables to, say, `__cedar::Default::"foo bar baz"`.  Is that right (or maybe I'm misreading)?

This seems potentially dangerous and confusing.  My inclination would be to restrict `__cedar::Default` to be an enum over three values (`__cedar::Default::"principal"`, `__cedar::Default::"action"`, `__cedar::Default::"resource"`), and to restrict the API accordingly, to make it impossible to construct requests that mix up the bindings of Cedar variables to these values.

Reply:
Good point. I also don't think we want users using the same default entity for both principal and resource, so I'll update the proposal to just disallow the second case. (It had questionable usefulness anyways.)